pkg/blockbuilder/blockbuilder.go

```diff
@@ -0,0 +1,509 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package blockbuilder
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/services"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/objstore"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kerr"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/plugin/kprom"
+
+	"github.com/grafana/mimir/pkg/storage/bucket"
+	"github.com/grafana/mimir/pkg/storage/ingest"
+	storagetsdb "github.com/grafana/mimir/pkg/storage/tsdb"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+type BlockBuilder struct {
+	services.Service
+
+	cfg         Config
+	logger      log.Logger
+	register    prometheus.Registerer
+	limits      *validation.Overrides
+	kafkaClient *kgo.Client
+	bucket      objstore.Bucket
+
+	parts []int32
```

Comment:
[nit] Let's make it clear. "Parts" means a different thing in english.

```suggestion
	partitionIDs []int32
```

Reply:
Should we even rename it to `ownedPartitionIDs` to be even more clear?